services/src/main/java/io/grpc/services/CallMetricRecorder.java

```diff
@@ -62,41 +101,118 @@ public static CallMetricRecorder getCurrent() {
     if (disabled) {
       return this;
     }
-    if (metrics.get() == null) {
+    if (requestCostMetrics.get() == null) {
       // The chance of race of creation of the map should be very small, so it should be fine
       // to create these maps that might be discarded.
-      metrics.compareAndSet(null, new ConcurrentHashMap<String, Double>());
+      requestCostMetrics.compareAndSet(null, new ConcurrentHashMap<String, Double>());
     }
-    metrics.get().put(name, value);
+    requestCostMetrics.get().put(name, value);
     return this;
   }
 
   /**
-   * Returns all save metric values. No more metric values will be recorded after this method is
-   * called. Calling this method multiple times returns the same collection of metric values.
+   * Records a call metric measurement for CPU utilization.
+   * If RPC has already finished, this method is no-op.
+   *
+   * <p>A latter record will overwrite its former name-sakes.
+   *
+   * @return this recorder object
+   * @since 1.47.0
+   */
+  public CallMetricRecorder recordCpuUtilizationMetric(double value) {
+    if (disabled) {
+      return this;
+    }
+    cpuUtilizationMetric = value;
+    return this;
+  }
+
+  /**
+   * Records a call metric measurement for memory utilization.
+   * If RPC has already finished, this method is no-op.
+   *
+   * <p>A latter record will overwrite its former name-sakes.
+   *
+   * @return this recorder object
+   * @since 1.47.0
+   */
+  public CallMetricRecorder recordMemoryUtilizationMetric(double value) {
+    if (disabled) {
+      return this;
+    }
+    memoryUtilizationMetric = value;
+    return this;
+  }
+
+
+  /**
+   * Returns all request cost metric values. No more metric values will be recorded after this
+   * method is called. Calling this method multiple times returns the same collection of metric
+   * values.
    *
    * @return a map containing all saved metric name-value pairs.
    */
   Map<String, Double> finalizeAndDump() {
     disabled = true;
-    Map<String, Double> savedMetrics = metrics.get();
+    Map<String, Double> savedMetrics = requestCostMetrics.get();
     if (savedMetrics == null) {
       return Collections.emptyMap();
     }
     return Collections.unmodifiableMap(savedMetrics);
   }
 
+  /**
+   * Returns all save metric values. No more metric values will be recorded after this method is
+   * called. Calling this method multiple times returns the same collection of metric values.
+   *
+   * @return a per-request ORCA reports containing all saved metrics.
+   */
+  CallMetricReport finalizeAndDump2() {
+    Map<String, Double> savedRequestCostMetrics = finalizeAndDump();
+    Map<String, Double> savedUtilizationMetrics = utilizationMetrics.get();
+    if (savedUtilizationMetrics == null) {
+      savedUtilizationMetrics = Collections.emptyMap();
+    }
+    return new CallMetricReport(cpuUtilizationMetric,
+        memoryUtilizationMetric, ImmutableMap.copyOf(savedRequestCostMetrics),
```

Comment:
The code has `disabled` so that it can avoid making the copy. Since this is per-RPC, I think we do want to avoid the re-creation overhead since it provides virtually no gain.

If this was a public API, we may choose to use Collections.unmodifiableMap. Since it is internal and the old code didn't, I think even that is unnecessary but I'm fine either way.